src/StoragelessSession/Http/SessionMiddleware.php

```diff
@@ -217,18 +218,32 @@ public function extractSessionContainer(Token $token = null) : Data
     /**
      * @param Data     $sessionContainer
      * @param Response $response
+     * @param Token    $originalToken
      *
      * @return Response
      *
      * @throws \InvalidArgumentException
      */
-    private function appendToken(Data $sessionContainer, Response $response) : Response
+    private function appendToken(Data $sessionContainer, Response $response, Token $originalToken = null) : Response
     {
-        if ($sessionContainer->isEmpty()) {
+        $empty = $sessionContainer->isEmpty();
+
+        if ($empty && null === $originalToken) {
```

Comment:
Note that this improvement is mainly to be more cache-friendly: varnish and similar reverse http proxies apply a "skip everything" policy to anything that contains a cookie (by default)